packages/@aws-cdk/core/lib/asset-staging.ts

```diff
@@ -12,6 +14,84 @@ export interface AssetStagingProps extends FingerprintOptions {
    * The source file or directory to copy from.
    */
   readonly sourcePath: string;
+
+  /**
+   * Bundle the asset by executing a command in a Docker container.
+   * The asset path will be mounted at `/asset-input`. The Docker
+   * container is responsible for putting content at `/asset-output`.
+   * The content at `/asset-output` will used as the final asset.
+   *
+   * @default - source is copied to staging directory
+   *
+   * @experimental
+   */
+  readonly bundling?: BundlingOptions;
+}
+
+/**
+ * Bundling options
+ *
+ * @experimental
+ */
+export interface BundlingOptions {
+  /**
+   * The Docker image where the command will run.
+   */
+  readonly image: BundlingDockerImage;
+
+  /**
+   * The command to run in the container.
+   *
+   * @example ['npm', 'install']
+   *
+   * @see https://docs.docker.com/engine/reference/run/
+   *
+   * @default - run the command defined in the image
+   */
+  readonly command?: string[];
+
+  /**
+   * Additional Docker volumes to mount.
+   *
+   * @default - no additional volumes are mounted
+   */
+  readonly volumes?: DockerVolume[];
+
+  /**
+   * The environment variables to pass to the container.
+   *
+   * @default - no environment variables.
+   */
+  readonly environment?: { [key: string]: string; };
+
+  /**
+   * Working directory inside the container.
+   *
+   * @default /asset-input
+   */
+  readonly workingDirectory?: string;
+
+  /**
+   * How to calculate the source hash for this asset.
+   *
+   * @default BUNDLE
+   */
+  readonly hashCalculation?: AssetHashCalculation;
```

Comment:
Feels like this belongs at the StagingOptions level since it interplays with the ability to specify a custom hash. 

We should also add CUSTOM here and require that if a custom hash is specified this option will either be undefined or set to CUSTOM.

Reply:
Initially also thought about having this on `AssetStagingProps` then I wrote as comment above `BUNDLE` *This only makes sense when bundling*, which made me move it. For the custom hash it is handled in `s3-assets` currently:

https://github.com/aws/aws-cdk/blob/6407535863c06d6d3ccfc2c3f2b59470d2d88993/packages/%40aws-cdk/aws-s3-assets/lib/asset.ts#L112

It only plays a role when calling `addFileAsset()` in `s3-assets`:

https://github.com/aws/aws-cdk/blob/6407535863c06d6d3ccfc2c3f2b59470d2d88993/packages/%40aws-cdk/aws-s3-assets/lib/asset.ts#L125-L129

So wouldn't `CUSTOM` be passing a hash to `Staging` and just giving it back?

Reply:
Feels like there's also something to do with the `extraHash`... at least update the JSDoc for the default

https://github.com/aws/aws-cdk/blob/6407535863c06d6d3ccfc2c3f2b59470d2d88993/packages/%40aws-cdk/core/lib/fs/options.ts#L55-L63

Reply:
I think bundling makes sense even if *custom bundling* is not configured. In that case the hash will be simply calculated on the .zip file. So I still feel it makes sense to move this from here. The bundling configuration doesn't "care" about hashes. It's just a set of build instructions.